Add ansible server configuration

bmispelon opened the pull request.

bmispelon committed:
Add ansible server configuration

bmispelon commented on ansible/README.md:
````diff
@@ -0,0 +1,66 @@
+To create a test file for development, you can copy the `hypha_secrets.env.example` one:
+```
+cp hypha_secrets.env.example hypha_secrets.env
+```
````
Not sure if that's the best way to manage secrets, but without the convenience of `heroku config` that's the best I could come up with.

Another option could be to use an ansible vault to encrypt the secrets, that way we could check the file in source control.

bmispelon commented on ansible/roles/hypha/templates/wtsh-hypha-web.service.jinja2:
```diff
@@ -0,0 +1,29 @@
+
+[Service]
+Restart=on-failure
+ExecStart=/usr/bin/docker run \
+    --name wtsh-hypha-web \
+    --pull never \
+    --rm \
+    --env-file /opt/hypha/wtsh-hypha-secrets.env \
+    --env PYTHONDONTWRITEBYTECODE=1 \
+    --env PYTHONUNBUFFERED=1 \
+    --env PROJECTS_ENABLED=1 \
+    --env ORG_SHORT_NAME=WT.SH \
+    --env ALLOWED_HOSTS={{ hypha_domain }} \
+    --env SECURE_SSL_REDIRECT=0 \
+    --env DATABASE_URL=postgres://hypha:hypha@wtsh-hypha-db/hypha \
+    --network wtsh-hypha \
+    --publish 127.0.0.1:9001:9001 \
+    ghcr.io/torchbox/wtsh-hypha:{{ hypha_docker_tag }}
```
I think this would also need a volume to persist uploaded files (`MEDIA_ROOT`) 🤔

bmispelon commented on ansible/roles/postgres/tasks/main.yml:
```diff
@@ -0,0 +1,24 @@
+- name: Create /opt/pg directory
+  ansible.builtin.file:
+    path: /opt/pg
+    state: directory
```
Not sure if I should restrict the `mode` here. Since everything runs as root I guess it doesn't matter much, right?

bmispelon commented on ansible/setup.yml:
```diff
@@ -0,0 +1,19 @@
+    - system
+    - docker
+    - postgres
+    - redis
```
The `wtsh-hypha` image currently hardcodes the `CACHES` settings to use the database backend, so technically `redis` is not used (yet).

bmispelon commented on ansible/setup.yml:
```diff
@@ -0,0 +1,19 @@
+    - hypha
+  vars:
+    # The tag of the docker image we want to deploy
+    hypha_docker_tag: wtsh-main
```
The current github action setup doesn't seem to publish a `latest` tag so I'm using the branch name instead.

bmispelon commented on ansible/roles/hypha/templates/wtsh-hypha-web.service.jinja2:
```diff
@@ -0,0 +1,29 @@
+Restart=on-failure
+ExecStart=/usr/bin/docker run \
+    --name wtsh-hypha-web \
+    --pull never \
```
The image is pulled by another task in the playbook so no need to pull here (the same applies to the other docker systemd unit files)

bmispelon commented on ansible/roles/hypha/templates/wtsh-hypha-web.service.jinja2:
```diff
@@ -0,0 +1,29 @@
+    --env ORG_SHORT_NAME=WT.SH \
+    --env ALLOWED_HOSTS={{ hypha_domain }} \
+    --env SECURE_SSL_REDIRECT=0 \
+    --env DATABASE_URL=postgres://hypha:hypha@wtsh-hypha-db/hypha \
```
The password is exposed here, but I think it's OK considering the postgres container doesn't expose any ports, right?

tomusher replied:
**question:** Why not specify all of these in the .env file?

bmispelon replied:
Good question. I was trying to distinguish between "public" env variables and "secret" ones, and keep the env file for the latter.

tomusher commented on ansible/roles/docker/tasks/main.yml:
```diff
@@ -0,0 +1,44 @@
+    state: started
+    enabled: yes
+
+- name: Create wtsh-hypha docker network
```
**suggestion:** Feels like this might belong in the `hypha` role? That way if we reuse the role, each instance gets their own network.

bmispelon replied:
I agree. I'd originally put it there because this network is shared between the three roles `postgres`, `redis` and `hypha` and the inter-dependency made it hard to put it in `hypha`.
But now that I've separated "setup" operations from "deploy" ones it's much easier to move the creation of the network over to the `hypha` role 👍🏻

tomusher commented on ansible/roles/hypha/tasks/main.yml:
```diff
@@ -0,0 +1,57 @@
+- name: Create /opt/hypha directory
+  ansible.builtin.file:
+    path: /opt/hypha
```
**suggestion (non-blocking):** Aware that this is probably bike shedding, this should probably be `/srv/hypha`

bmispelon replied:
Happy to pick your color for the bikeshed 👍🏻

tomusher commented on ansible/roles/hypha/tasks/main.yml:
**suggestion:** This role could take a 'project name' or 'environment' variable so it could be used multiple times to deploy different environments on the same host.

bmispelon replied:
I've created a `defaults/main.yml` file with a bunch of variables to make it easier to customise/reuse the `hypha` role.
In practice I'm not sure the role is 100% reusable since the `postgres` and `redis` roles both depend on the existence of the `wtsh-hypha` docker network.

tomusher commented on ansible/roles/hypha/tasks/main.yml:
```diff
@@ -0,0 +1,57 @@
+    state: link
+  notify: Restart nginx
+
+- name: Pull hypha docker image
```
**suggestion:** This may be personal preference, but I don't think pulling and running the image should be the responsibility of the playbook. I'd normally separate out provisioning and deployment processes so that:
- You don't need to run the playbook/need Ansible for deployment
- You can make provisioning/infrastructure changes without affecting the current deployment

Happy to hear arguments to the contrary though.

bmispelon replied:
That's a good point. Because I've been working on setting up a server from scratch I've been conflating the two steps but they should really be separate.

I've moved all "deploy" steps into a separate "deploy.yml" playbook for now. I'm still not sure what strategy I should adopt for automated deployment, I'd like to talk through options with you if that's possible.

tomusher commented on ansible/roles/hypha/tasks/main.yml:
```diff
@@ -0,0 +1,57 @@
+    enabled: true
+    daemon_reload: true
+
+- name: Run management commands
```
**suggestion**: As above, these feel like deployment steps rather than provisioning steps.

tomusher commented on ansible/roles/postgres/tasks/main.yml:
```diff
@@ -0,0 +1,23 @@
+- name: Create /opt/pg directory
+  ansible.builtin.file:
+    path: /opt/pg
```
**suggestion:** Again, this is not a big deal, but perhaps putting this in `/srv` makes sense too.

tomusher commented on ansible/roles/postgres/tasks/main.yml:
```diff
@@ -0,0 +1,23 @@
+    path: /opt/pg
+    state: directory
+
+- name: Pull postgres docker image
```
**suggestion:** (applies to Redis too) You could let Docker pull the image (with the 'missing' pull policy) which means you don't have to have Ansible pull it. At the moment you need to update the image tag twice (once in the role, once in the template), which risks one being missed when upgrading.

bmispelon replied:
Good point. I've moved the tag name to a role-level variable (with a default), but I also removed the explicit pull. Originally I'd tried to mimic the way it was done for the `hypha` image but the extra step doesn't make much sense here.

tomusher reviewed (commented):
This is looking good to me!

I've added a few comments inline but have a few overall thoughts to discuss.

**Mixing provisioning with deployment** - The playbook currently tries to both provision the server and deploy the app. I consider these separate concerns and generally think this tends to make it harder/slower to deploy and makes it harder to reason about what's going to happen when you want to make an infrastructure change and *not* deploy, or vice versa.

**Systemd to run docker** - By using systemd with `docker run` your systemd unit ends up monitoring the docker client process, rather than the container ([more info here](https://github.com/moby/moby/issues/6791)).
This means that the unit generally won't realise if the container dies.
If we want to address this (and I'm not saying we have to - I have prod apps running this way and it hasn't been a huge problem), then there's a few options:
- Use something like podman ([which does support systemd natively](https://docs.podman.io/en/latest/markdown/podman-systemd.unit.5.html)) over Docker.
- Let Docker manage its own processes, e.g. using docker compose

**Running as root** - It's an annoying amount of extra work but it would be good if we could run the image rootless. [Docker supports this](https://docs.docker.com/engine/security/rootless/), as does [Podman](https://github.com/containers/podman/blob/main/docs/tutorials/rootless_tutorial.md).

bmispelon commented:
Thanks for the review, I've addressed most of your comments/suggestions except for these:

1) real systemd integration: i see your point, but that seems like a big change to make at this point. How about I create a "SHOULD" ticket for this on Jira and I can address it if there's time at the end of the sprint
2) Similarly, a rootless/podman setup would be great, but I'm worried about not having enough time to complete that. What about another "SHOULD" ticket for it?

Some questions left to answer:

1) The current setup needs to have 777 permissions on the `media` directory, otherwise the deploy management commands throw an error about file permissions. There's probably something I'm misunderstanding about docker volume and host file permissions here, I could use some help 🤔 
2) Decide on a strategy for automated deployment (this will also inform/impact how we deal with secrets)

bmispelon committed:
Add github worklfow for deploying

bmispelon commented:
I've tidied up the git history and pushed the following features:

- The ansible playbook is now only for setting up the server, it doesn't handle deploying.
- Instead, a deploy.sh script is generated on the server, along with a `deploy` user whose only permissions are to run that script (via `sudo`)
- I've added a `deploy.yml` github workflow that will ssh to the appropriate server (prod or staging) and run that script (I've also set up the corresponding two environments in this project's settings: `prod` and `staging`).

bmispelon committed:
Disabled deploying to prod for now (there is no prod)

tomusher commented on ansible/roles/hypha/tasks/main.yml:
```diff
@@ -0,0 +1,58 @@
+    dest: "{{ hypha_root_dir }}/hypha.env"
+    force: false
+
+- name: Create docker network {{ hypha_docker_network }}
```
**Question:** Should this network be created before the db service starts?

bmispelon replied:
Err, yes. I must have missed that while moving the tasks around (it doesn't help that the restart task seems to be run in a sort of async way and won't error out if the docker container fails to start).
I've moved the docker network creation nearer the top, thanks 👍🏻

tomusher commented on ansible/roles/hypha/defaults/main.yml:
```diff
@@ -0,0 +1,29 @@
+hypha_postgres_version: 16
+
+# The name of the system user that can deploy the site (used in github action)
+hypha_deploy_user: deploy
```
**suggestion:** As you're using a separate deploy script for each environment, perhaps this user should use `{{ app_name }}` in its name so you don't have to explicitly override it for each environment?

Alternatively, a single deploy script/user for both environments would probably be fine.

bmispelon replied:
Overriding `hypha_deploy_user` should only really be needed when deploying two environments on the same server, otherwise using the default is safe.
However the failure mode of reusing the same deploy user for two environments on the same host is pretty bad, since you'd just end up silently deploying whichever environment was last set up no matter whether you're trying to deploy prod or staging.
So I've made taken your suggestion and made the deploy username dynamic (based on `app_name`0 to avoid that failure mode.

tomusher reviewed (approved):
This looks good to me.

One general point (which I'd consider non-blocking, maybe something for the backlog) is that I think we should be deploying specific versioned image tags (likely just the commit hash is fine) rather than using the branch name.

This would make it easier to see what version is deployed and would make rollbacks easier.

bmispelon commented:
> One general point (which I'd consider non-blocking, maybe something for the backlog) is that I think we should be deploying specific versioned image tags (likely just the commit hash is fine) rather than using the branch name.

Ah I see, that's what you used `SSH_ORIGINAL_COMMAND` in the `authorized_keys` for in your original example.

That shouldn't be too hard to add. I'll see if I have some time left at the end of this sprint (it'll require tweaking both the deploy script and the github action).

Could you explain how it would help to see which version has been deployed? I don't know where GitHub surfaces that information? 🤔

tomusher commented:
> Could you explain how it would help to see which version has been deployed? I don't know where GitHub surfaces that information? 🤔

You'd be able to see the current tag in the unit file/when insepcting the docker container which is useful when checking exactly what commit/version is running. When it's always 'latest' you need to check the image hash and then reference that back to the registry/actions which can be cumbersome.
